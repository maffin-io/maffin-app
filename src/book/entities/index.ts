export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
