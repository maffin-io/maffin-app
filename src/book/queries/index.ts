export { getInvestments } from './getInvestments';
export { getMainCurrency } from './getMainCurrency';
export { getEarliestDate } from './getEarliestDate';
export { getSplits } from './getSplits';
export { getLatestTxs } from './getLatestTxs';
