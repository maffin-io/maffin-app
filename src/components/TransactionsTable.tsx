import React from 'react';
import { ColumnDef, CellContext, Row } from '@tanstack/react-table';
import classNames from 'classnames';
import Link from 'next/link';
import { FaArrowUp, FaArrowDown } from 'react-icons/fa';
import { BiEdit, BiXCircle } from 'react-icons/bi';

import TransactionFormButton from '@/components/buttons/TransactionFormButton';
import Table from '@/components/Table';
import Tooltip from '@/components/Tooltip';
import Money from '@/book/Money';
import {
  Account,
  Commodity,
  Split,
} from '@/book/entities';

export type TransactionsTableProps = {
  splits: Split[],
  accounts: Account[],
};

export default function TransactionsTable({
  splits,
  accounts,
}: TransactionsTableProps): JSX.Element {
  columns[2].cell = FromToAccountPartial(accounts);

  return (
    <Table<Split>
      columns={columns}
      data={splits}
    />
  );
}

const columns: ColumnDef<Split>[] = [
  {
    header: 'Date',
    id: 'date',
    enableSorting: false,
    accessorFn: (row: Split) => row.transaction.date.toMillis(),
    cell: ({ row }) => (
      <Tooltip
        text={row.original.transaction.guid}
      >
        <span>
          {row.original.transaction.date.toISODate()}
        </span>
      </Tooltip>
    ),
  },
  {
    header: 'Description',
    enableSorting: false,
    accessorFn: (row: Split) => row.transaction.description,
  },
  {
    header: 'From/To',
    enableSorting: false,
  },
  {
    accessorKey: 'value',
    header: 'Amount',
    enableSorting: false,
    cell: ({ row }) => {
      let value = new Money(row.original.quantity, row.original.account.commodity.mnemonic);

      if (row.original.account.type === 'INCOME') {
        value = value.multiply(-1);
      }
      return (
        <span
          className={
            classNames({
              'text-green-300': value.toNumber() > 0,
              'text-red-300': value.toNumber() < 0,
            }, 'flex items-center')
          }
        >
          {value.toNumber() > 0 && <FaArrowUp className="text-xs mr-1" />}
          {value.toNumber() < 0 && <FaArrowDown className="text-xs mr-1" />}
          {value.format()}
        </span>
      );
    },
  },
  {
    header: 'Total',
    enableSorting: false,
    cell: ({ row, table }) => {
      const { rows } = table.getCoreRowModel();
      const currentRow = rows.findIndex(
        (otherRow: Row<Split>) => otherRow.original.guid === row.original.guid,
      );
      const nextRows = rows.slice(currentRow, rows.length + 1);
      const totalPreviousSplits = nextRows.reverse().reduce(
        (total: Money, otherRow: Row<Split>) => {
          if (otherRow.original.account.type === 'INCOME') {
            return total.add(
              new Money(-otherRow.original.quantity, otherRow.original.account.commodity.mnemonic),
            );
          }
          return total.add(
            new Money(otherRow.original.quantity, otherRow.original.account.commodity.mnemonic),
          );
        },
        new Money(0, row.original.account.commodity.mnemonic),
      );

      return (
        <span>
          {totalPreviousSplits.format()}
        </span>
      );
    },
  },
  {
    header: 'Actions',
    enableSorting: false,
    cell: ({ row }) => (
      <>
        <TransactionFormButton
          action="update"
          defaultValues={
            {
              ...row.original.transaction,
              date: row.original.transaction.date.toISODate() as string,
              fk_currency: row.original.transaction.currency as Commodity,
            }
          }
          className="link"
        >
          <BiEdit className="flex" />
        </TransactionFormButton>
        <TransactionFormButton
          action="delete"
          defaultValues={
            {
              ...row.original.transaction,
              date: row.original.transaction.date.toISODate() as string,
              fk_currency: row.original.transaction.currency as Commodity,
            }
          }
          className="link"
        >
          <BiXCircle className="flex" />
        </TransactionFormButton>
      </>
    ),
  },
];

function FromToAccountPartial(
  accounts: Account[],
) {
  return function FromToAccount({ row }: CellContext<Split, unknown>): JSX.Element {
    const { splits } = row.original.transaction;
    const otherSplits = splits.filter(split => split.account.guid !== row.original.account.guid);

    return (
      <ul>
        { otherSplits.map(split => {
          const account = accounts.find(a => a.guid === split.account.guid);

          return (
            <li key={split.guid}>
              <Link
                href={`/dashboard/accounts/${split.account.guid}`}
                className={classNames('badge hover:text-slate-300', {
                  'bg-green-500/20 text-green-300': split.account.type === 'INCOME',
                  'bg-red-500/20 text-red-300': split.account.type === 'EXPENSE',
                  'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(split.account.type),
                  'bg-orange-500/20 text-orange-300': split.account.type === 'LIABILITY',
                  'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(split.account.type),
                })}
              >
                { account?.path }
              </Link>
            </li>
          );
        })}
      </ul>
    );
  };
}
