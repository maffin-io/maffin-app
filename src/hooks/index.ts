export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
