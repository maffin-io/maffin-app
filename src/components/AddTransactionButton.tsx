import React from 'react';
import { BiPlusCircle } from 'react-icons/bi';

import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Account } from '@/book/entities';
import Modal from '@/components/Modal';
import Tooltip from '@/components/Tooltip';

type AddTransactionButtonProps = {
  account: Account,
};

export default function AddTransactionButton(
  {
    account,
  }: AddTransactionButtonProps,
): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const disabled = ['INCOME', 'EXPENSE'].includes(account.type);

  return (
    <>
      <Modal
        title={`${account.path} transaction`}
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <TransactionForm
          onSave={() => { setIsModalOpen(false); }}
          account={account}
        />
      </Modal>
      <Tooltip
        text="Add transactions from accounts that are not INCOME/EXPENSE/STOCK/MUTUAL"
        show={disabled}
      >
        <button
          type="button"
          className="btn-primary"
          disabled={disabled}
          onClick={() => setIsModalOpen(!isModalOpen)}
        >
          <BiPlusCircle className="mr-1" />
          Add Transaction
        </button>
      </Tooltip>
    </>
  );
}
