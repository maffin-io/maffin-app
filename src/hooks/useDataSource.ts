import React from 'react';
import { DataSource } from 'typeorm';
import initSqlJs from 'sql.js';
import { mutate } from 'swr';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/Stocker';
import useBookStorage from '@/hooks/useBookStorage';
import { importBook as importGnucashBook } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { isStaging } from '@/helpers/env';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        await loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  mutate('/state/save', true, { revalidate: false });
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  await storage.save(rawBook);
  mutate('/state/save', false, { revalidate: false });
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const rawBook = await importGnucashBook(parsedData);
  await DATASOURCE.sqljsManager.loadDatabase(rawBook);

  DATASOURCE.options.extra.queryClient.refetchQueries({
    type: 'all',
  });

  await loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  if (!isStaging()) {
    try {
      // We have to await because we use Price.upsert. Once we correct this,
      // saving Price already updates the local cache so we don't need await
      await insertTodayPrices();
    } catch (e) {
      console.warn(`Retrieving live prices failed ${e}`);
    }
  }
}
