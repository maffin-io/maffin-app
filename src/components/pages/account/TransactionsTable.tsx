import React from 'react';
import {
  ColumnDef,
  CellContext,
  PaginationState,
} from '@tanstack/react-table';
import classNames from 'classnames';
import Link from 'next/link';
import { FaArrowUp, FaArrowDown } from 'react-icons/fa';
import { BiEdit, BiXCircle } from 'react-icons/bi';
import { Tooltip } from 'react-tooltip';
import { useQuery, UseQueryResult } from '@tanstack/react-query';

import FormButton from '@/components/buttons/FormButton';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import Table from '@/components/Table';
import Money from '@/book/Money';
import {
  Account,
  Split,
  Transaction,
} from '@/book/entities';
import { useSplitsCount, useSplitsPagination } from '@/hooks/api';
import type { Commodity } from '@/book/entities';
import { accountColorCode } from '@/helpers/classNames';
import fetcher from '@/hooks/api/fetcher';

export type TransactionsTableProps = {
  account: Account,
};

export default function TransactionsTable({
  account,
}: TransactionsTableProps): JSX.Element {
  const [{ pageIndex, pageSize }, setPagination] = React.useState<PaginationState>({
    pageIndex: 0,
    pageSize: 10,
  });
  const { data: splitsCount } = useSplitsCount(account.guid);
  const { data: splits } = useSplitsPagination(account.guid, { pageIndex, pageSize });

  columns[3].cell = AmountPartial(account);
  columns[4].cell = TotalPartial(account);

  const pagination = React.useMemo(
    () => ({
      pageIndex,
      pageSize,
    }),
    [pageIndex, pageSize],
  );

  return (
    <Table<Split>
      id="transactions-table"
      columns={columns}
      data={splits || []}
      showPagination
      onPaginationChange={setPagination}
      pageCount={Math.ceil((splitsCount || 0) / pageSize)}
      state={{
        pagination,
      }}
      manualPagination
    />
  );
}

const columns: ColumnDef<Split>[] = [
  {
    header: 'Date',
    id: 'date',
    enableSorting: false,
    accessorFn: (row: Split) => row.transaction.date.toMillis(),
    cell: ({ row }) => (
      <>
        <span
          data-tooltip-id={row.original.txId}
          data-tooltip-content={row.original.txId}
        >
          {row.original.transaction.date.toISODate()}
        </span>
        <Tooltip clickable className="tooltip" id={row.original.txId} />
      </>
    ),
  },
  {
    header: 'Description',
    enableSorting: false,
    cell: DescriptionCell,
  },
  {
    header: 'From/To',
    enableSorting: false,
    cell: FromToAccountCell,
  },
  {
    accessorKey: 'value',
    header: 'Amount',
    enableSorting: false,
  },
  {
    header: 'Total',
    enableSorting: false,
  },
  {
    header: 'Actions',
    enableSorting: false,
    cell: ActionsCell,
  },
];

function useTransaction(guid: string): UseQueryResult<Transaction> {
  const queryKey = [...Transaction.CACHE_KEY, guid];
  return useQuery({
    queryKey,
    queryFn: fetcher(
      () => Transaction.findOne({
        where: { guid },
        relations: {
          splits: {
            fk_account: true,
          },
        },
      }),
      queryKey,
    ),
  });
}

function DescriptionCell({ row }: CellContext<Split, unknown>): JSX.Element {
  const { data: tx } = useTransaction(row.original.txId);
  return (
    <span>
      {tx?.description}
    </span>
  );
}

function FromToAccountCell({ row }: CellContext<Split, unknown>): JSX.Element {
  const { data: tx } = useTransaction(row.original.txId);

  const otherSplits = tx?.splits.filter(
    split => split.accountId !== row.original.accountId,
  ) || [];

  return (
    <ul>
      { otherSplits.map(split => {
        const { account } = split;

        return (
          <li key={split.guid}>
            <Link
              href={`/dashboard/accounts/${split.accountId}`}
              className={accountColorCode(account, 'badge mb-0.5 hover:text-slate-300')}
            >
              { account?.path }
            </Link>
          </li>
        );
      })}
    </ul>
  );
}

function AmountPartial(
  account: Account,
) {
  return function AmountCell({ row }: CellContext<Split, unknown>): JSX.Element {
    let value = new Money(row.original.quantity, account.commodity.mnemonic || '');

    if (account?.type === 'INCOME') {
      value = value.multiply(-1);
    }
    return (
      <span
        className={
          classNames({
            'amount-positive': value.toNumber() > 0,
            'amount-negative': value.toNumber() < 0,
          }, 'flex items-center')
        }
      >
        {value.toNumber() > 0 && <FaArrowUp className="text-xs mr-1" />}
        {value.toNumber() < 0 && <FaArrowDown className="text-xs mr-1" />}
        {value.format()}
      </span>
    );
  };
}

function TotalPartial(
  account: Account,
) {
  return function TotalCell({ row }: CellContext<Split, unknown>): JSX.Element {
    return (
      <span>
        {new Money(row.original.balance, account.commodity.mnemonic).format()}
      </span>
    );
  };
}

function ActionsCell({ row }: CellContext<Split, unknown>): JSX.Element {
  const { data: tx } = useTransaction(row.original.txId);

  if (!tx || tx.guid !== row.original.txId) {
    return <span />;
  }

  const originalSplit = tx.splits.find(split => split.guid === row.original.guid) as Split;
  const defaultValues = {
    ...tx,
    date: tx.date.toISODate() as string,
    fk_currency: tx.fk_currency as Commodity,
    // This is hacky but if we pass the Split
    // class to the form, then we have reference errors as when
    // we update the form, it also updates the defaultValues
    // which means formState.isDirty is not triggered properly
    splits: [
      {
        ...originalSplit,
        value: originalSplit.value,
        quantity: originalSplit.quantity,
      },
      ...tx.splits.filter(split => split.guid !== originalSplit.guid).map(split => ({
        ...split,
        value: split.value,
        quantity: split.quantity,
      } as Split)),
    ] as Split[],
  };
  return (
    <>
      <FormButton
        id="edit-tx"
        modalTitle="Edit transaction"
        buttonContent={<BiEdit className="flex" />}
        className="link"
      >
        <TransactionForm
          action="update"
          defaultValues={defaultValues}
        />
      </FormButton>
      <FormButton
        id="delete-tx"
        modalTitle="Confirm you want to remove this transaction"
        buttonContent={<BiXCircle className="flex" />}
        className="link"
      >
        <TransactionForm
          action="delete"
          defaultValues={defaultValues}
        />
      </FormButton>
    </>
  );
}
