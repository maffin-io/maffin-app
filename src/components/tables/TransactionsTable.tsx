import React from 'react';
import {
  ColumnDef,
  CellContext,
  PaginationState,
} from '@tanstack/react-table';
import classNames from 'classnames';
import Link from 'next/link';
import { FaArrowUp, FaArrowDown } from 'react-icons/fa';
import { BiEdit, BiXCircle } from 'react-icons/bi';
import { DateTime } from 'luxon';

import FormButton from '@/components/buttons/FormButton';
import { Tooltip } from '@/components/tooltips';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import SearchBox from '@/components/SearchBox';
import Table from '@/components/tables/Table';
import Money from '@/book/Money';
import {
  Account,
  Split,
} from '@/book/entities';
import {
  useAccounts,
  useSplitsCount,
  useSplitsPagination,
  useTransaction,
} from '@/hooks/api';
import { accountColorCode } from '@/helpers/classNames';

export type TransactionsTableProps = {
  account: Account,
};

export default function TransactionsTable({
  account,
}: TransactionsTableProps): React.JSX.Element {
  const [search, setSearch] = React.useState('');
  const [{ pageIndex, pageSize }, setPagination] = React.useState<PaginationState>({
    pageIndex: 0,
    pageSize: 10,
  });
  const { data: splitsCount } = useSplitsCount(account.guid, search);
  const { data: splits } = useSplitsPagination(
    account.guid,
    { pageSize, pageIndex },
    search,
  );

  columns[3].cell = AmountPartial(account);
  columns[4].cell = BalancePartial(account);

  const pagination = React.useMemo(
    () => ({
      pageIndex,
      pageSize,
    }),
    [pageIndex, pageSize],
  );

  return (
    <>
      <SearchBox onChange={setSearch} />
      <Table<Split>
        id="transactions-table"
        columns={columns}
        data={splits || []}
        showPagination
        onPaginationChange={setPagination}
        pageCount={Math.ceil((splitsCount || 0) / pageSize)}
        state={{
          pagination,
        }}
        manualPagination
      />
    </>
  );
}

const columns: ColumnDef<Split>[] = [
  {
    header: 'Date',
    id: 'date',
    enableSorting: false,
    accessorFn: (row: Split) => row.transaction.date.toMillis(),
    cell: ({ row }) => (
      <>
        <span
          data-tooltip-id={row.original.txId}
          data-tooltip-content={row.original.txId}
        >
          {row.original.transaction.date.toLocaleString(DateTime.DATE_SHORT)}
        </span>
        <Tooltip clickable id={row.original.txId} />
      </>
    ),
  },
  {
    header: 'Description',
    enableSorting: false,
    cell: DescriptionCell,
  },
  {
    header: 'From/To',
    enableSorting: false,
    cell: FromToAccountCell,
  },
  {
    accessorKey: 'value',
    header: 'Amount',
    enableSorting: false,
  },
  {
    header: 'Balance',
    enableSorting: false,
  },
  {
    header: 'Actions',
    enableSorting: false,
    cell: ActionsCell,
  },
];

function DescriptionCell({ row }: CellContext<Split, unknown>): React.JSX.Element {
  const { data: tx } = useTransaction({ guid: row.original.txId });
  return (
    <span>
      {tx?.description}
    </span>
  );
}

function FromToAccountCell({ row }: CellContext<Split, unknown>): React.JSX.Element {
  const { data: tx } = useTransaction({ guid: row.original.txId });
  const { data: accounts } = useAccounts();

  const otherSplits = tx?.splits.filter(
    split => split.accountId !== row.original.accountId,
  ) || [];

  return (
    <ul>
      { otherSplits.map(split => {
        const account = accounts?.find(a => a.guid === split.accountId) as Account;

        return (
          <li key={split.guid}>
            <Link
              href={`/dashboard/accounts/${split.accountId}`}
              className={accountColorCode(account, 'badge mb-0.5 hover:text-slate-300')}
            >
              { account?.path }
            </Link>
          </li>
        );
      })}
    </ul>
  );
}

function AmountPartial(
  account: Account,
) {
  return function AmountCell({ row }: CellContext<Split, unknown>): React.JSX.Element {
    let value = new Money(row.original.quantity, account.commodity.mnemonic || '');

    if (account?.type === 'INCOME') {
      value = value.multiply(-1);
    }
    return (
      <span
        className={
          classNames({
            'text-success': value.toNumber() > 0,
            'text-danger': value.toNumber() < 0,
          }, 'flex items-center')
        }
      >
        {value.toNumber() > 0 && <FaArrowUp className="text-xs mr-1" />}
        {value.toNumber() < 0 && <FaArrowDown className="text-xs mr-1" />}
        {value.format()}
      </span>
    );
  };
}

function BalancePartial(
  account: Account,
) {
  return function BalanceCell({ row }: CellContext<Split, unknown>): React.JSX.Element {
    return (
      <span>
        {new Money(row.original.balance, account.commodity.mnemonic).format()}
      </span>
    );
  };
}

function ActionsCell({ row }: CellContext<Split, unknown>): React.JSX.Element {
  const { data: tx } = useTransaction({ guid: row.original.txId });

  if (!tx || tx.guid !== row.original.txId) {
    return <span />;
  }

  return (
    <>
      <FormButton
        id="edit-tx"
        modalTitle="Edit transaction"
        buttonContent={<BiEdit className="flex" />}
        className="text-left text-cyan-700 hover:text-cyan-600"
      >
        <TransactionForm
          action="update"
          guid={row.original.txId}
          onSave={() => tx.queryClient?.invalidateQueries({
            queryKey: [...Split.CACHE_KEY, row.original.accountId],
          })}
        />
      </FormButton>
      <FormButton
        id="delete-tx"
        modalTitle="Confirm you want to remove this transaction"
        buttonContent={<BiXCircle className="flex" />}
        className="text-left text-cyan-700 hover:text-cyan-600"
      >
        <TransactionForm
          action="delete"
          guid={row.original.txId}
        />
      </FormButton>
    </>
  );
}
