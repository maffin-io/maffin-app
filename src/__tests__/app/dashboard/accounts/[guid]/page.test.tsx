import React from 'react';
import {
  waitFor,
  render,
  screen,
} from '@testing-library/react';
import type { DataSource } from 'typeorm';

import AccountPage from '@/app/dashboard/accounts/[guid]/page';
import { TransactionsTableProps } from '@/components/TransactionsTable';
import { AddTransactionButtonProps } from '@/components/AddTransactionButton';
import { Account } from '@/book/entities';
import * as queries from '@/book/queries';
import * as dataSourceHooks from '@/hooks/useDataSource';

jest.mock('@/components/AddTransactionButton', () => {
  function AddTransactionButton(props: AddTransactionButtonProps) {
    return (
      <div className="AddTransactionButton">
        <span>{JSON.stringify(props)}</span>
      </div>
    );
  }

  return AddTransactionButton;
});

jest.mock('@/components/TransactionsTable', () => {
  function TransactionsTable(props: TransactionsTableProps) {
    return (
      <div className="TransactionsTable">
        <span>{JSON.stringify(props)}</span>
      </div>
    );
  }

  return TransactionsTable;
});

jest.mock('@/book/queries', () => ({
  __esModule: true,
  ...jest.requireActual('@/book/queries'),
}));

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));
// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/navigation'), 'useRouter');

describe('AccountPage', () => {
  let mockRouterPush: jest.Mock;

  beforeEach(() => {
    mockRouterPush = jest.fn();
    useRouter.mockImplementation(() => ({
      push: mockRouterPush,
    }));
    jest.spyOn(queries, 'getAccountsWithPath').mockResolvedValue([
      {
        guid: 'guid',
        path: 'path',
        type: 'TYPE',
      } as Account,
    ]);
  });

  it('displays loading while account is null', () => {
    jest.spyOn(Account, 'findOneBy').mockResolvedValue(null);
    const { container } = render(<AccountPage params={{ guid: 'guid' }} />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('returns 404 when account not found', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);
    jest.spyOn(queries, 'getAccountsWithPath').mockResolvedValue([]);

    render(
      <AccountPage params={{ guid: 'guid' }} />,
    );

    await waitFor(() => expect(mockRouterPush).toHaveBeenCalledWith('/404'));
  });

  it('renders as expected with account', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);

    const { container } = render(
      <AccountPage params={{ guid: 'guid' }} />,
    );

    await screen.findByText(/accountId/);
    expect(container).toMatchSnapshot();
  });
});
