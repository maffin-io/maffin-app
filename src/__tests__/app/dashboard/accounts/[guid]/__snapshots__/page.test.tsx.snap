// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountPage displays loading while account is null 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`AccountPage renders as expected with account 1`] = `
<div>
  <div
    class="grid grid-cols-12 items-center pb-4"
  >
    <span
      class="col-span-10 text-xl font-medium"
    >
      path
       
      account
    </span>
    <div
      class="col-span-2 col-end-13 justify-self-end"
    >
      <div
        class="AddTransactionButton"
      >
        <span>
          {"account":{"guid":"guid","path":"path","type":"TYPE"}}
        </span>
      </div>
    </div>
  </div>
  <div
    class="TransactionsTable"
  >
    <span>
      {"accountId":"guid","accounts":[{"guid":"guid","path":"path","type":"TYPE"}]}
    </span>
  </div>
</div>
`;
